### PR TITLE
Support Sprockets 4.0!

### PR DESCRIPTION
**Pull request by hmdne**
Noting things of importance:
* Improved (robust!) source map support - no more source map comments in a middle of a concatenated file
* Caching: it's blazingly fast
* Potential support for transpiling ES6. Once Opal becomes ES6-reliant, it will just warrant a small switch for everything to keep working
* You can now name your templates just ".erb" and everything will work correctly

Things that will need upgrade once this is deployed (or before - backward compatibility is assured): opal-haml, opal-erubi (I already have those and I will reference this PR).

Things that will NOT need upgrade: opal-rspec, opal

Things I'm not sure about, but will probably need some minor work: opal-rails

### CODE REVIEW

**Comment by hmdne** [comment]
Regarding the break:

    opal-sprockets was resolved to 0.5.0.1.0.4.0, which depends on
          sprockets (~> 4.0) was resolved to 4.0.2, which depends on
            ruby (>= 2.5.0)

I think this patch is partially compatible with sprockets 3.7 if that's desired.

**Comment by elia** [comment]
This is amazing! I'll review it as soon as I can 💪 ✨ 🎉

**Comment by hmdne** [comment]
Regarding opal-rails, Rails changed their Sprockets handling method if Sprockets 4.0 is loaded, it now needs a file app/assets/config/manifest.js. I will try to migrate it, but I have very little experience with Rails. Do note, that nothing changes for other deployments, only Rails is affected.

**Comment by hmdne** [comment]
I found a problem that dependencies may not work the way they should. While updating my code, the asset (which is now just one asset, it doesn't expand anymore) compiles only the main file without its dependencies and a server restart is required. I will create a test for that instance and try to debug it more.

File: spec/sprockets/server_spec.rb [hunk 1]
```diff
@@ -39,6 +39,29 @@ def app
...
+    get '/assets/require_tree_test.js'
+    does_include.(last_response.body, [1,2], [3])
+
+    File.write(__dir__+"/../fixtures/required_tree_test/required_file3.rb", "p 3")
+
+    asset = app.sprockets['require_tree_test', accept: "application/javascript"]
+    expect(asset).to be_truthy
+    does_include.(asset.to_s, [1,2,3])
+    get '/assets/require_tree_test.js'
+    does_include.(last_response.body, [1,2,3])
```
Comment (hmdne) [inline spec/sprockets/server_spec.rb:60]: Here it fails

**Comment by hmdne** [comment]
Testing this on my application. It works flawlessly if opal-parser isn't included. And without collapsing everything loads very fast - it only recompiles single files from the source code. But if opal-parser is included it breaks and it breaks quite hard - sometimes it even embeds source maps as if they were sources. Maybe I will try later with increasing the Sprockets cache to more than 1k keys and hopefully this would help.

**Comment by elia** [comment]
That's great! I need to try it out as well, feel free to update the travis config to avoid unsupported rubies like 2.3 and 2.4 so that the PR becomes green

File: lib/opal/sprockets/path_reader.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module Opal
   module Sprockets
-
+    # Note: I don't think it's used anymore.
```
Comment (elia) [inline lib/opal/sprockets/path_reader.rb:3]: It's ok to remove the file altogether if that's the case

**Comment by tobymao** [comment]
the caching support is exciting! any update when this will get in?

**Comment by hmdne** [comment]
@tobymao I will finalize the patch in a few days. In the meantime you can do

`gem 'opal-sprockets', github: 'hmdne/opal-sprockets'`

etc. for opal-rails, opal-haml.

**Comment by elia** [comment]
I merged this with a few adjustments in #23, thanks for the great work!

**Pull request closed without merging**